### ISSUE TITLE
Int overflow in hpa causing incorrect replica count

### ISSUE THREAD

**Issue by sheepster1** [kubernetes/kubernetes#126892]
### What happened?

The setup:
I am using keda with the prometheus scaler. The query I am using, returns the lag in the message queue i am using, and the threshold is set to `0.1`.

What happened:
The lag was increasing for a long time, and the replica count reached the max setting as expected. Everything was running fine for some time. When the lag value reached `214,748,364` hpa decided to reduce the replicas from the max limit to `1`.

What I think is the problem:
When the lag passes `214,748,364`, the calculation [here](https://github.com/kubernetes/kubernetes/blob/7b80cdb66a390f225d23cd612950144e3a39d1ae/pkg/controller/podautoscaler/replica_calculator.go#L278) divides by the threshold `0.1` and it passes the max int32 value. causing hpa to scale to the minimum value, 1.
It also seems like a lot of other places in this file cast a 64 bit float to a 32 bit int. Should there maybe be a check everywhere this is done?

### What did you expect to happen?

I expected the replica count to stay at the max value. Or alternatively, get an error that we have reached the max value for an external metric value

### How can we reproduce it (as minimally and precisely as possible)?

Use an external metric, and set it above `214,748,364` with a threshold of `0.1`.

### Kubernetes version

<details>

```console
$ kubectl version
# paste output here
```
1.29
</details>

### Cloud provider

<details>
aws eks
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by vaibhav2107** [comment]
As PR is still open, 
/remove-lifecycle stale

**Comment by adrianmoisey** [comment]
> Use an external metric, and set it above `214,748,364` with a threshold of `0.1`.

What do you mean by `threshold`? Is that the target?

**Comment by sheepster1** [comment]
@adrianmoisey yes, I am pretty sure that the "threshold" in keda is translated into "target"

**Comment by adrianmoisey** [comment]
I'm not sure I can reproduce the problem, take a look at this:
```go
package main

import (
	"fmt"
	"math"
)

func main() {
	// usageRatio := float64(usage) / float64(targetUsage)
	// https://github.com/kubernetes/kubernetes/blob/7b80cdb66a390f225d23cd612950144e3a39d1ae/pkg/controller/podautoscaler/replica_calculator.go#L260
	below := float64(214748364) / float64(0.1)
	above := float64(214748365) / float64(0.1)

	// replicaCount = int32(math.Ceil(usageRatio * float64(readyPodCount)))
	// https://github.com/kubernetes/kubernetes/blob/7b80cdb66a390f225d23cd612950144e3a39d1ae/pkg/controller/podautoscaler/replica_calculator.go#L278
	fmt.Println(int32(math.Ceil(below * float64(int64(100)))))
	fmt.Println(int32(math.Ceil(above * float64(int64(100)))))
}
```

```console
$ go run .
2147483647
2147483647
```

**Comment by sheepster1** [comment]
I'm am running the exact code snippet you sent and I'm getting:
```
 ⇒  go run .
 -2147483648
 -2147483648
```

**Comment by sheepster1** [comment]
I'm not very familiar with go to be honest. But casting a 64 bit variable to a 32 bit variable is a problem in other languages.
And it seems to have happened on my production env consistently until we capped the external metric value artificially.

Regarding this test https://github.com/kubernetes/kubernetes/issues/126892#issuecomment-2585649993, what go version are you using to run this? I'll try match mine exactly to see is I can replicate that

**Comment by adrianmoisey** [comment]
Taking a look at https://go.dev/play/p/GUMaImIkOnT, it seems like you're right. 
I'm unsure what's happening on my local, but I guess we can reliably reproduce this now.

This makes me think that https://github.com/kubernetes/kubernetes/pull/127050 is right, it was my local testing that didn't agree.

**Comment by sheepster1** [comment]
I'm assuming you are running on an apple silicone machine.
Tested it now on a Mac and managed to replicate your output and then my outpost with these containers:
```
 docker run --rm -it --platform linux/amd64 golang:latest # outputs negative values
 docker run --rm -it golang:latest # outputs positive values
```
I guess it's a go thing on different architectures..

**Comment by omerap12** [comment]
Related PR ( waiting to be merged ): #127050 
/remove-lifecycle stale

**Comment by omerd-cyera** [comment]
Great to see #127050 was merged! Thanks for everyone involved.
@omerap12 Are you planning to go forward with the other cases in this file like you mentioned [here](https://github.com/kubernetes/kubernetes/pull/126979#issuecomment-2320328068)?
To be honest, I'm not sure which of these unsafe casts are the cause for the original bug, so I don't know if this issue is resolved or not.

**Comment by omerap12** [comment]
> Great to see [#127050](https://github.com/kubernetes/kubernetes/pull/127050) was merged! Thanks for everyone involved. [@omerap12](https://github.com/omerap12) Are you planning to go forward with the other cases in this file like you mentioned [here](https://github.com/kubernetes/kubernetes/pull/126979#issuecomment-2320328068)? To be honest, I'm not sure which of these unsafe casts are the cause for the original bug, so I don't know if this issue is resolved or not.

Yeah, I re-opened it and assigned a reviewer. thanks for pinging :)
